### PR TITLE
NavigateTo command (part of #441)

### PR DESCRIPTION
**Pull request by helen-dikareva**
/cc @AlexanderMoskovkin @VasilyStrelyaev @AndreyBelym

### CODE REVIEW

File: src/client/runner/automation/playback/navigate-to.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+export default class NavigateToAutomation {
+    constructor (url) {
+        this.url = url;
+    }
+
+    run () {
+        hammerhead.navigateTo(this.url);
+
+        return delay(NAVIGATION_DELAY);
```
Comment (inikulin) [inline src/client/runner/automation/playback/navigate-to.js]: Do we really need this delay? Wouldn't page load barrier handle this?

File: src/test-run/commands/index.js [hunk 2]
```diff
@@ -188,6 +188,23 @@ export class TypeTextCommand extends Assignable {
...
+
+        this.type = TYPE.navigateTo;
+        this.url  = null;
+
+        this._assignFrom(obj, true);
+    }
+
+    _getAssignableProperties () {
+        return [
+            { name: 'url', type: nonEmptyStringArgument, required: true }
```
Comment (inikulin) [inline src/test-run/commands/index.js]: I think we need to validate protocol as well (it should be `http`, `https`, relative protocol - `//` or relative URL). BTW, I guess we should allow relative urls, I'm not sure if current `navigateTo` implementation in hammerhead supports relative URLs - we need to check this and create a separate issue for this one.

**Comment by inikulin** [comment]
\r-

File: src/test-run/commands/index.js (hunk 2 above)
Comment (helen-dikareva) [inline src/test-run/commands/index.js]: Hammerhead supports relative URLs now.

Comment (inikulin) [inline src/test-run/commands/index.js]: Ok, then we need tests for 3 scenarios: 
- URL with protocol 
- URL with relative protocol 
- relative URL

Comment (AlexanderMoskovkin) [inline src/test-run/commands/index.js]: @inikulin do you mean we need a `NavigateToCommand` `url` validator and we should raise an error if the `url` arguement is not valid?

And what the reason to test different urls here? We just call the `hammerhead.navigateTo`function , I suppose it should be tested in `hammerhead`?

Comment (inikulin) [inline src/test-run/commands/index.js]: > We just call the hammerhead.navigateTofunction , I suppose it should be tested in hammerhead

I have no idea. Also, I have no idea what will happen if we pass incorrect URL that's why I want tests which will describe this behavior (and help us avoid regressions).

File: test/functional/fixtures/api/raw/navigate-to/test.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+var errorInEachBrowserContains = require('../../../../assertion-helper.js').errorInEachBrowserContains;
+
+
+describe('[Raw API] Navigate to action', function () {
+    it('Should make navigation to another page', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/raw/navigate-to/test.js]:
```
    it('Should navigate to another page', function () {
```

**Comment by VasilyStrelyaev** [comment]
\r-

File: test/functional/fixtures/api/raw/navigate-to/test.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+var errorInEachBrowserContains = require('../../../../assertion-helper.js').errorInEachBrowserContains;
+
+
+describe('[Raw API] Navigate to action', function () {
+    it('Should navigate to page using url that contains protocol', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/raw/navigate-to/test.js]:
```
    it('Should navigate to a page using a url that contains a protocol', function () {
```

File: test/functional/fixtures/api/raw/navigate-to/test.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        return runTests('./testcafe-fixtures/navigate-to.testcafe', 'Navigate to page using url that contains protocol', { shouldFail: true })
+            .catch(function (errs) {
+                errorInEachBrowserContains(errs, 'Click raised on page after navigation', 0);
+            });
+    });
+
+    it('Should navigate to page using protocol-relative url', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/raw/navigate-to/test.js]:
```
    it('Should navigate to a page using a protocol-relative url', function () {
```

File: test/functional/fixtures/api/raw/navigate-to/test.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        return runTests('./testcafe-fixtures/navigate-to.testcafe', 'Navigate to page using protocol-relative url', { shouldFail: true })
+            .catch(function (errs) {
+                errorInEachBrowserContains(errs, 'Click raised on page after navigation', 0);
+            });
+    });
+
+    it('Should navigate to page using relative url', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/raw/navigate-to/test.js]:
```
    it('Should navigate to a page using a relative url', function () {
```

File: test/functional/fixtures/api/raw/navigate-to/test.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        return runTests('./testcafe-fixtures/navigate-to.testcafe', 'Navigate to page using relative url', { shouldFail: true })
+            .catch(function (errs) {
+                errorInEachBrowserContains(errs, 'Click raised on page after navigation', 0);
+            });
+    });
+
+    it('Should navigate to 404-page using non-existent url', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/raw/navigate-to/test.js]:
```
    it('Should navigate to a 404-page using a non-existent url', function () {
```

File: test/functional/fixtures/api/raw/navigate-to/test.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        return runTests('./testcafe-fixtures/navigate-to.testcafe', 'Navigate to page using non-existent url');
+    });
+
+    it('Should navigate to 404-page using incorrect url', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/raw/navigate-to/test.js]:
```
    it('Should navigate to the 404-page using an incorrect url', function () {
```

File: test/functional/fixtures/api/raw/navigate-to/testcafe-fixtures/navigate-to.testcafe [hunk 9]
```diff
@@ -0,0 +1,67 @@
+{
+    "fixtures": [
+        {
+            "name": "Navigate to",
+            "pageUrl": "http://localhost:3000/api/raw/navigate-to/pages/index.html",
+            "tests": [
+                {
+                    "name": "Navigate to page using url that contains protocol",
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/raw/navigate-to/testcafe-fixtures/navigate-to.testcafe]:
```
                    "name": "Navigate to a page using a url that contains a protocol",
```

File: test/functional/fixtures/api/raw/navigate-to/testcafe-fixtures/navigate-to.testcafe [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+                            "url": "http://localhost:3000/api/raw/navigate-to/pages/navigation.html"
+                        },
+                        {
+                            "type": "click",
+                            "selector": "#button"
+                        }
+                    ]
+                },
+                {
+                    "name": "Navigate to page using protocol-relative url",
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/raw/navigate-to/testcafe-fixtures/navigate-to.testcafe]:
```
                    "name": "Navigate to a page using a protocol-relative url",
```

File: test/functional/fixtures/api/raw/navigate-to/testcafe-fixtures/navigate-to.testcafe [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+                            "url": "//localhost:3000/api/raw/navigate-to/pages/navigation.html"
+                        },
+                        {
+                            "type": "click",
+                            "selector": "#button"
+                        }
+                    ]
+                },
+                {
+                    "name": "Navigate to page using relative url",
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/raw/navigate-to/testcafe-fixtures/navigate-to.testcafe]:
```
                    "name": "Navigate to a page using a relative url",
```

File: test/functional/fixtures/api/raw/navigate-to/testcafe-fixtures/navigate-to.testcafe [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+                            "url": "navigation.html"
+                        },
+                        {
+                            "type": "click",
+                            "selector": "#button"
+                        }
+                    ]
+                },
+                {
+                    "name": "Navigate to page using non-existent url",
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/raw/navigate-to/testcafe-fixtures/navigate-to.testcafe]:
```
                    "name": "Navigate to a page using a non-existent url",
```

File: test/functional/fixtures/api/raw/navigate-to/testcafe-fixtures/navigate-to.testcafe [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+                    "commands": [
+                        {
+                            "type": "navigate-to",
+                            "url": "http://localhost:3000/api/raw/navigate-to/pages/non-existent.html"
+                        }
+                    ]
+                },
+                {
+                    "name": "Navigate to page using incorrect url",
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/raw/navigate-to/testcafe-fixtures/navigate-to.testcafe]:
```
                    "name": "Navigate to a page using an incorrect url",
```

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by helen-dikareva** [comment]
FPR

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by AndreyBelym** [comment]
lgtm

File: src/client/runner/api/actions.js [hunk 14]
```diff
@@ -807,13 +807,10 @@ export function waitFor (event, timeout) {
...
 
 export function navigateTo (url) {
-    var NAVIGATION_DELAY = 1000;
-
     stepIterator.asyncAction(function (iteratorCallback) {
         hammerhead.navigateTo(url);
 
-        //NOTE: give browser some time to navigate
-        nativeMethods.setTimeout.call(window, iteratorCallback, NAVIGATION_DELAY);
+        iteratorCallback();
```
Comment (AlexanderMoskovkin) [inline src/client/runner/api/actions.js]: Sure we don't need the `NAVIGATION_DELAY` in the new api. But should we remove it from the legacy api? 
@inikulin as I remember you are the one who implemented the `navigateTo` action in legacy api. Do you have thoughts why we were need this delay?

**Comment by AlexanderMoskovkin** [comment]
lgtm in others
@inikulin please, review too

File: src/client/runner/api/actions.js (hunk 14 above)
Comment (inikulin) [inline src/client/runner/api/actions.js]: Legacy API is in maintance state now. Try to keep everything as is there

**Comment by helen-dikareva** [comment]
@testcafe-build-bot /retest

**Comment by helen-dikareva** [comment]
FPR

File: src/errors/test-run/type.js [hunk 15]
```diff
@@ -20,6 +20,7 @@ export default {
     actionPositiveIntegerArgumentError: 'actionPositiveIntegerArgumentError',
     actionSelectorTypeError:            'actionSelectorTypeError',
     actionAdditionalSelectorTypeError:  'actionAdditionalSelectorTypeError',
+    actionUnsupportedProtocolError:     'actionUnsupportedProtocolError',
```
Comment (inikulin) [inline src/errors/test-run/type.js]: `actionUnsupportedUrlProtocolError

File: src/test-run/commands/index.js [hunk 16]
```diff
@@ -63,14 +68,23 @@ function positiveIntegerArgument (name, val) {
...
     if (type !== 'string')
-        throw new ActionStringArgumentError(option, type);
+        throw new ActionStringArgumentError(name, type);
 
     if (!val.length)
-        throw new ActionStringArgumentError(option);
+        throw new ActionStringArgumentError(name);
+}
+
+function navigateToArgument (name, val) {
```
Comment (inikulin) [inline src/test-run/commands/index.js]: `urlArgument`

File: src/test-run/commands/index.js [hunk 17]
```diff
@@ -63,14 +68,23 @@ function positiveIntegerArgument (name, val) {
...
+    nonEmptyStringArgument(name, val);
+
+    var url = val.trim();
+
+    if (PROTOCOL_RE.test(url) && !SUPPORTED_PROTOCOL_RE.test(url))
```
Comment (inikulin) [inline src/test-run/commands/index.js]: Why we need `PROTOCOL_RE` test?

Comment (inikulin) [inline src/test-run/commands/index.js]: Got it. Can we do this in context of single regular expression?

File: test/functional/fixtures/api/raw/navigate-to/testcafe-fixtures/navigate-to.testcafe [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+                            "selector": "#button"
+                        }
+                    ]
+                },
+                {
+                    "name": "Navigate to a page using a relative url",
+                    "commands": [
+                        {
+                            "type": "navigate-to",
+                            "url": "navigation.html"
```
Comment (inikulin) [inline test/functional/fixtures/api/raw/navigate-to/testcafe-fixtures/navigate-to.testcafe]: Can we use something like `./navigation.html` instead?

**Comment by inikulin** [comment]
\r-

File: src/test-run/commands/index.js (hunk 17 above)
Comment (helen-dikareva) [inline src/test-run/commands/index.js]: I think single regex will be too complicated.
I suggest rewrite it:
```
function navigateToUrlArgument (name, val) {
    nonEmptyStringArgument(name, val);

    var url      = val.trim();
    var protocol = PROTOCOL_RE.match(url);

    if (protocol && !SUPPORTED_PROTOCOL_RE.test(protocol))
        throw new ActionUnsupportedProtocolError(name, protocol);
}
```
And use `protocol` in errror's text:
`The <code>${err.argumentName}</code> argument specifies a URL that uses an unsupported <code>err.protocol</code> protocol. Only HTTP and HTTPS are supported, as well as protocol-relative and relative URLs.`
What do you think @inikulin ?

Comment (inikulin) [inline src/test-run/commands/index.js]: lgtm

**Comment by helen-dikareva** [comment]
FPR

**Comment by inikulin** [comment]
lgtm

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Pull request merged**